src/runtime/security/kinit_context.cpp

```diff
@@ -0,0 +1,306 @@
+    KRB5_RETURN_NOT_OK(krb5_kt_resolve(_krb5_context, FLAGS_krb5_keytab, &_keytab),
+                       "couldn't resolve keytab file");
+
+    // acquire credential cache handle
+    KRB5_RETURN_NOT_OK(krb5_cc_default(_krb5_context, &_ccache),
+                       "couldn't acquire credential cache handle");
+
+    // initialize credential cache
+    KRB5_RETURN_NOT_OK(krb5_cc_initialize(_krb5_context, _ccache, _principal),
+                       "initialize credential cache failed");
+
+    // allocate a new initial credential options structure
+    KRB5_RETURN_NOT_OK(krb5_get_init_creds_opt_alloc(_krb5_context, &_opt),
+                       "alloc get_init_creds_opt structure failed");
+
+    // get and schedule to renew credentials from KDC and store it into _ccache
+    RETURN_NOT_OK(get_credentials());
+    schedule_renew_credentials();
+
+    return error_s::ok();
+}
+
+void kinit_context_impl::init_krb5_ctx()
+{
+    static std::once_flag once;
+    std::call_once(once, [&]() {
+        int64_t err = krb5_init_context(&_krb5_context);
+        if (err != 0) {
+            dassert_f(false, "init kerberos context failed, with kerberos  error_code = {}", err);
+        }
+    });
+}
+
+error_s kinit_context_impl::parse_username_from_principal()
+{
+    // Attention: here we just assume the length of username must be little than 1024
+    const uint16_t BUF_LEN = 1024;
+    char buf[BUF_LEN];
+    krb5_error_code err = krb5_aname_to_localname(_krb5_context, _principal, sizeof(buf), buf);
+
+    // KRB5_LNAME_NOTRANS means no translation available for requested principal
+    if (err == KRB5_LNAME_NOTRANS) {
+        if (_principal->length > 0) {
+            int cnt = 0;
+            while (cnt < _principal->length) {
+                std::string tname;
+                tname.assign((const char *)_principal->data[cnt].data,
+                             _principal->data[cnt].length);
+                if (!_user_name.empty()) {
+                    _user_name += '/';
+                }
+                _user_name += tname;
+                cnt++;
+            }
+            return error_s::ok();
+        }
+        return error_s::make(ERR_KRB5_INTERNAL, "parse username from principal failed");
+    }
+
+    // KRB5_CONFIG_NOTENUFSPACE means BUF_LEN is not enough
+    if (err == KRB5_CONFIG_NOTENUFSPACE) {
+        return error_s::make(ERR_KRB5_INTERNAL, fmt::format("username is larger than {}", BUF_LEN));
+    }
+    KRB5_RETURN_NOT_OK(err, "krb5 parse aname to localname failed");
+
+    if (strlen(buf) <= 0) {
+        return error_s::make(ERR_KRB5_INTERNAL, "empty username");
+    }
+
+    _user_name.assign((const char *)buf);
+    return error_s::ok();
+}
+
+error_s kinit_context_impl::get_credentials()
+{
+    krb5_creds creds;
+    error_s err = error_s::ok();
+
+    // get initial credentials using a key table
+    // Notice: the contents of a krb5_creds structure need to be freed by ourselves
```

Comment:
Dose this comment mean that we should call function `krb5_free_cred_contents` to free krb5_creds structure?

Reply:
yes